genie-web/src/main/java/com/netflix/genie/web/rpc/grpc/services/impl/v4/GRpcJobTerminationServiceImpl.java

```diff
@@ -0,0 +1,120 @@
+ *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *     See the License for the specific language governing permissions and
+ *     limitations under the License.
+ *
+ */
+package com.netflix.genie.web.rpc.grpc.services.impl.v4;
+
+import com.google.common.collect.Maps;
+import com.netflix.genie.common.exceptions.GenieException;
+import com.netflix.genie.common.exceptions.GenieServerException;
+import com.netflix.genie.proto.JobKillListenerRequest;
+import com.netflix.genie.proto.JobKillListenerResponse;
+import com.netflix.genie.proto.JobTerminationServiceGrpc;
+import com.netflix.genie.web.rpc.grpc.interceptors.SimpleLoggingInterceptor;
+import com.netflix.genie.web.services.JobKillServiceV4;
+import com.netflix.genie.web.services.JobSearchService;
+import io.grpc.stub.StreamObserver;
+import lombok.extern.slf4j.Slf4j;
+import net.devh.springboot.autoconfigure.grpc.server.GrpcService;
+
+import javax.validation.constraints.NotBlank;
+import java.util.Map;
+
+/**
+ * Extension of {@link JobTerminationServiceGrpc.JobTerminationServiceImplBase}
+ * to provide functionality for killing jobs running on agents.
+ * TODO Register with HeartBeatService to listen for agent stream to become inactive and clean it up.
+ *
+ * @author standon
+ * @since 4.0.0
+ */
+@GrpcService(
+    value = JobTerminationServiceGrpc.class,
+    interceptors = {
+        SimpleLoggingInterceptor.class,
+    }
+)
+@Slf4j
+public class GRpcJobTerminationServiceImpl
+    extends JobTerminationServiceGrpc.JobTerminationServiceImplBase
+    implements JobKillServiceV4 {
+
+    private final Map<String, StreamObserver<JobKillListenerResponse>> parkedJobKillResponseObservers =
+        Maps.newConcurrentMap();
+    private final JobSearchService jobSearchService;
+
+    /**
+     * Constructor.
+     *
+     * @param jobSearchService Job search service
+     */
+    public GRpcJobTerminationServiceImpl(final JobSearchService jobSearchService) {
+        this.jobSearchService = jobSearchService;
+    }
+
+    /**
+     * Register to be notified when a kill request for the job is received.
+     *
+     * @param request          Request to register for getting notified when server gets a job kill request.
+     * @param responseObserver The response observer
+     */
+    @Override
+    public void registerForKillNotification(
+        final JobKillListenerRequest request,
+        final StreamObserver<JobKillListenerResponse> responseObserver
+    ) {
+        parkedJobKillResponseObservers.put(request.getJobId(), responseObserver);
+    }
+
+    @Override
+    public void killJob(final @NotBlank(message = "No job id entered. Unable to kill job.") String jobId,
+                        final @NotBlank(message = "No reason provided.") String reason) throws GenieException {
+
+        final StreamObserver<JobKillListenerResponse> responseObserver =
+            cleanUpAndGetParkedJobKillResponseObserver(jobId);
```

Comment:
The more I think about it, I think we should remove cleanup of response obsevers completely from here and have the cleanup done only when a disconnect happens. Thoughts?